sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuAvroScan.scala

```diff
@@ -745,18 +745,31 @@ class GpuMultiFileCloudAvroPartitionReader(
             } else {
               estBlocksSize
             }
+            val optHmb = if (readDataSchema.nonEmpty) {
+              Some(HostMemoryBuffer.allocate(headerSize + estSizeToRead))
+            } else None
             // Allocate the buffer for the header and blocks for a batch
-            closeOnExcept(HostMemoryBuffer.allocate(headerSize + estSizeToRead)) { hmb =>
-              val out = new HostMemoryOutputStream(hmb)
-              // Write the header to the output stream
-              AvroFileWriter(out).writeHeader(reader.header)
+            closeOnExcept(optHmb) { _ =>
+              val optOut = optHmb.map { hmb =>
+                val out = new HostMemoryOutputStream(hmb)
+                // Write the header to the output stream
+                AvroFileWriter(out).writeHeader(reader.header)
+                out
+              }
               // Read the block data to the output stream
               var batchRowsNum: Int = 0
+              var batchSize: Long = 0
               var hasNextBlock = true
               do {
-                reader.readNextRawBlock(out)
+                if (optOut.nonEmpty) {
```

Comment:
NIT: optOut.map(xx).getOrElse(xxx)

Reply:
done

Reply:
Not a good idea, `map` is usually used to convert a to b, but `readNextRawBlock` and `skipCurrentBlock` do not return anything, meaning you are trying to convert out to nothing(aka unit).

Reply:
ok, rolled back.